src/data/logs.ts

```diff
@@ -0,0 +1,264 @@
+ */
+
+type LogsVolumeQueryOptions<T extends DataQuery> = {
+  targets: T[];
+  range: TimeRange;
+};
+
+const MILLISECOND = 1;
+const SECOND = 1000 * MILLISECOND;
+const MINUTE = 60 * SECOND;
+const HOUR = 60 * MINUTE;
+const DAY = 24 * HOUR;
+
+const LogLevelColor = {
+  [LogLevel.critical]: colors[7],
+  [LogLevel.warning]: colors[1],
+  [LogLevel.error]: colors[4],
+  [LogLevel.info]: colors[0],
+  [LogLevel.debug]: colors[5],
+  [LogLevel.trace]: colors[2],
+  [LogLevel.unknown]: getThemeColor('#8e8e8e', '#bdc4cd'),
+};
+
+function getThemeColor(dark: string, light: string): string {
+  return config.bootData.user.lightTheme ? light : dark;
+}
+
+/**
+ * Creates an observable, which makes requests to get logs volume and aggregates results.
+ */
+export function queryLogsVolume<TQuery extends DataQuery, TOptions extends DataSourceJsonData>(
+  datasource: DataSourceApi<TQuery, TOptions>,
+  logsVolumeRequest: DataQueryRequest<TQuery>,
+  options: LogsVolumeQueryOptions<TQuery>
+): Observable<DataQueryResponse> {
+  return new Observable((observer) => {
+    let rawLogsVolume: DataFrame[] = [];
+    observer.next({
+      state: LoadingState.Loading,
+      error: undefined,
+      data: [],
+    });
+
+    const queryResponse = datasource.query(logsVolumeRequest);
+    const queryObservable = isObservable(queryResponse) ? queryResponse : from(queryResponse);
+
+    const subscription = queryObservable.subscribe({
+      complete: () => {
+        const aggregatedLogsVolume = aggregateRawLogsVolume(rawLogsVolume);
+        if (aggregatedLogsVolume[0]) {
+          aggregatedLogsVolume[0].meta = {
+            custom: {
+              targets: options.targets,
+              absoluteRange: { from: options.range.from.valueOf(), to: options.range.to.valueOf() },
+            },
+          };
+        }
+        observer.next({
+          state: LoadingState.Done,
+          error: undefined,
+          data: aggregatedLogsVolume,
+        });
+        observer.complete();
+      },
+      next: (dataQueryResponse: DataQueryResponse) => {
+        const { error } = dataQueryResponse;
+        if (error !== undefined) {
+          observer.next({
+            state: LoadingState.Error,
+            error,
+            data: [],
+          });
+          observer.error(error);
+        } else {
+          rawLogsVolume = rawLogsVolume.concat(dataQueryResponse.data.map(toDataFrame));
+        }
+      },
+      error: (error: DataQueryError) => {
+        observer.next({
+          state: LoadingState.Error,
+          error: error,
+          data: [],
+        });
+        observer.error(error);
+      },
+    });
+    return () => {
+      subscription?.unsubscribe();
+    };
+  });
+}
+
+/**
+ * Take multiple data frames, sum up values and group by level.
+ * Return a list of data frames, each representing single level.
+ */
+function aggregateRawLogsVolume(rawLogsVolume: DataFrame[]): DataFrame[] {
+  if (rawLogsVolume.length !== 1) {
```

Comment:
just to clarify a bit: we already getting a single DataFrame here which looks something like this:

```json
{
    "fields": [
        {
            "name": "time",
            "type": "time",
            "typeInfo": {
                "frame": "time.Time"
            },
            "config": {},
            "values": [
                1680140003000,
                1680140004000,
                1680140027000
            ],
            "entities": {}
        },
        {
            "name": "emerg",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "fatal",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "alert",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "crit",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "critical",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "warn",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "warning",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "err",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "eror",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "error",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "info",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                1,
                1
            ],
            "entities": {}
        },
        {
            "name": "information",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "informational",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "notice",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "dbug",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "debug",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                1,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "trace",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        },
        {
            "name": "unknown",
            "type": "number",
            "typeInfo": {
                "frame": "uint64"
            },
            "config": {},
            "values": [
                0,
                0,
                0
            ],
            "entities": {}
        }
    ],
    "length": 3
}
```